packages/block-editor/src/hooks/style.js

```diff
@@ -159,13 +278,13 @@ export function addSaveProps(
 		const skipSerialization = getBlockSupport( blockType, indicator );
 
 		if ( skipSerialization === true ) {
-			style = omit( style, path );
+			style = omitStyle( style, path );
```

Comment:
Can `path` also be an array, or is it guaranteed to be a string?

Reply:
`path` can be a string, an array, an array of strings, and an array of arrays.

Reply:
I looked at the original code and couldn't fully understand why there was a need for a special treatment. However, from what I understand, this isn't something that's coming from this PR, so I believe it makes sense that we handle it in a different PR. Or am I misunderstanding you? What do you think?

Reply:
I don't have full context on the original implementation around `skipPaths` but when you check `skipSerializationPathsEdit` and `skipSerializationPathsSave`, their path values are all single element arrays. 

I believe they were designed that way specifically to align with the fact that each block support (border, color, spacing, typography etc) corresponds with a single top-level path within the style object.

That leads me to conclude that there wasn't the intent to omit multiple style paths in a single call here, [as proposed](https://github.com/WordPress/gutenberg/pull/45014/files#r996979744).

The initial call to omit a style occurs when an entire block support has its serialization skipped via a simple `true` value for its `__experiementalSkipSerialization` flag. For example, all `typography` styles. The next call is only to omit a single feature for a block support e.g. only skip text decoration but still serialize all other typography supports.

In the latter case, we want `style.typography.textDecoration` to be removed but the `style.typography` to remain.

I could well be misunderstanding your point @jsnajdr but hopefully the above helps clarify things some.

Reply:
> That leads me to conclude that there wasn't the intent to omit multiple style paths in a single call

Then why would `skipPaths` values be arrays? If the intent is to always omit only a single path, the `skipPaths` would be a `Record< string, string >`, like:
```js
skipPaths = {
  'spacing': 'spacing.blockGap'
}
```
but in fact it is a `Record< string, string[] >`, like:
```js
skipPaths = {
  'spacing': [ 'spacing.blockGap' ]
}
```
which IMO clearly shows that there can be potentially multiple paths to skip. The `skipPaths` structure is _designed_ to support that. And indeed a call like:
```js
paths = skipPaths[ 'spacing' ];
style = omit( style, paths );
```
would remove them all.

But the second `omit` call, the `omit( style, [ [ ...path, feature ] ] )`, uses `path` with different semantics. For our spacing example, that could lead to calling `omit` as:
```js
omit( style, [ [ 'spacing.blockGap', 'blockGap' ] ] )
```
which is a nonsense combination, it won't omit anything.

The Gallery block has spacing support defined as:
```
supports: {
  spacing: {
    /* ... */
    __experimentalSkipSerialization: [ "blockGap" ],
  }
}
```
So, when calling `addSaveProps` on such a Gallery block, the `spacing.blockGap` style won't be removed although it should be?

> However, from what I understand, this isn't something that's coming from this PR, so I believe it makes sense that we handle it in a different PR.

@tyxla It seems that the pre-existing code has a bug, caused by incorrect usage of `omit`. That seems very relevant to the current PR, and should get some attention. Otherwise the PR is kind of "garbage in, garbage out", not very valuable. Also, the PR doesn't modify anything else but `addSaveProps`, which is a reason why it's not interesting to land it separately without addressing issues.

Reply:
> It seems that the pre-existing code has a bug, caused by incorrect usage of omit

I think you might be right about a bug, but maybe not with the use of `omit`. 

It appears to be an issue with the check for skipping serialization or the config for skipping the blockGap support in `skipSerializationPathsSave`.

The `__experimentalSkipSerialization` flags are meant to  be a `boolean` or `array`. The block gap support looks like it was trying to always skip its serialization from being saved into block markup, likely due to that fact it was injected server-side. 

The indicator used to retrieve the block support config, when processing that block gap entry from `skipSerializationPathsSave`, was simply `spacing`, which would return that entire supports config `object`. 

I'd need to test that further, but it appears the block gap support wouldn't be being skipped for save at all, as that skip serialization value wouldn't be `true` or an `array`.

cc/ @andrewserong for your thoughts and additional background here.


> Then why would `skipPaths` values be arrays?

Good point. Looking back through the file history here it looks like they were arrays due to some of the typography styles originally not being included under a single `typography` banner like they are now.

They could probably be changed to strings not arrays now.


> But the second `omit` call, the `omit( style, [ [ ...path, feature ] ] )`, uses `path` with different semantics. For our spacing example, that could lead to calling `omit` as:
>
> `omit( style, [ [ 'spacing.blockGap', 'blockGap' ] ] )`

A call as suggested here wouldn't occur as the skip serialization retrieved from the block supports wouldn't be `true` or an `array`.

> So, when calling addSaveProps on such a Gallery block, the spacing.blockGap style won't be removed although it should be?

In this case, the `blockGap` style is being removed both in the edit and save contexts due to the normal `spacing` block support config in the `skipPaths`. When processing the skip paths, the indicator for the spacing support, `${ SPACING_SUPPORT_KEY }.__experimentalSkipSerialization`, would return the array `[ "blockGap" ]`. Which would then mean a call to `omit( style, [ [ 'spacing', 'blockGap' ] ] )`.

Reply:
> Otherwise the PR is kind of "garbage in, garbage out", not very valuable. Also, the PR doesn't modify anything else but addSaveProps, which is a reason why it's not interesting to land it separately without addressing issues.

Keeping the scope for this PR purely to the replacement of the Lodash `omit` function sounds like a win. I don't think it hurts to move the discussion on the logic contained in `addSaveProps` to its own dedicated PR rather than hijack this PR further.

Happy to leave that call to @tyxla though 🙂

Reply:
Now when we reached some agreement that there is indeed something wrong with the underlying code, and it's on someone's radar, my objections are no longer so strong 🙂

Reply:
Thanks, everyone for the thoughts and research put into this discussion.

@aaronrobertshaw basically said it here:

> Keeping the scope for this PR purely to the replacement of the Lodash omit function sounds like a win. I don't think it hurts to move the discussion on the logic contained in addSaveProps to its own dedicated PR rather than hijack this PR further.

This aligns with my sentiment, and I do think we should try our best to keep changes focused on one thing at a time. What if there were 1000 bugs in this component? Should we fix them all in a single PR? I don't think so. That doesn't mean that this shouldn't be addressed in a follow-up, though. 

I do feel like @aaronrobertshaw, @andrewserong, and @ramonjd are better equipped to address this one, but I'm happy to help with reviews and testing.

Any objections?